Add Automatic Directional Navigation Graph Generation

jbuehler23 opened the pull request:
## Objective

Resolves #21661 

Adds automatic directional navigation graph generation based on UI node positions and sizes, eliminating the need for tedious manual graph construction in dynamic UIs.

## Solution

Implements a spatial navigation algorithm that automatically computes the nearest neighbor in each compass direction for UI elements, while respecting any manually-defined edges.

### Features

- **Automatic edge generation**: Finds the best neighbor in each of 8 compass directions based on distance, alignment, and overlap
- **Manual override support**: Manual edges always take precedence over auto-generated ones
- **Configurable**: `AutoNavigationConfig` resource allows tuning alignment requirements, distance limits, and preference weighting
- **Opt-in**: Entities must have `AutoDirectionalNavigation` component added to use, and therefore **not** a breaking change
- **Generic**: Core algorithm works with any `Vec2` position/size data, not just `bevy_ui`

### Implementation

**New Components & Resources** (`bevy_input_focus/src/directional_navigation.rs`):
- `AutoDirectionalNavigation` - Marker component to enable auto-navigation
- `AutoNavigationConfig` - Configuration resource with settings:
  - `min_alignment_factor`: Minimum perpendicular overlap (0.0-1.0) required for cardinal directions
  - `max_search_distance`: Optional distance limit for connections
  - `prefer_aligned`: Whether to strongly prefer well-aligned nodes

**Core Algorithm**:
```rust
pub fn auto_generate_navigation_edges(
    nav_map: &mut DirectionalNavigationMap,
    nodes: &[(Entity, Vec2, Vec2)],  // (entity, center_pos, size)
    config: &AutoNavigationConfig,
)
```

For each node and each direction:
1. Filter candidates that are actually in that direction (cone-based check)
2. Calculate overlap factor for cardinal directions (horizontal overlap for N/S, vertical for E/W)
3. Score candidates based on:
   - Distance (closer is better)
   - Alignment with direction vector (more aligned is better)
   - Overlap factor (must meet minimum threshold)
4. Select the best-scoring candidate as the neighbor

**Scoring Formula**:
```
score = distance + alignment_penalty
where alignment_penalty = (1.0 - alignment) * distance * 2.0
```

This makes misaligned nodes significantly less attractive while still considering distance.

### Usage

**Before (manual)**:
```rust
// Must manually specify all connections
for row in 0..N_ROWS {
    let entities_in_row: Vec<Entity> = (0..N_COLS)
        .map(|col| button_entities.get(&(row, col)).unwrap())
        .copied()
        .collect();
    directional_nav_map.add_looping_edges(&entities_in_row, CompassOctant::East);
}

// Repeat for columns...
for col in 0..N_COLS {
    let entities_in_column: Vec<Entity> = (0..N_ROWS)
        .map(|row| button_entities.get(&(row, col)).unwrap())
        .copied()
        .collect();
    directional_nav_map.add_edges(&entities_in_column, CompassOctant::South);
}
```

**After**:
```rust
// Just add the `AutoDirectionalNavigation` component!
commands.spawn((
    Button,
    Node { /* ... */ },
    AutoDirectionalNavigation::default(),
    // ... other components
));
```

## Testing
- Added new example: `auto_directional_navigation`
- Ran existing `directional_navigation`

## Showcase

### New Example: `auto_directional_navigation`

Demonstrates automatic navigation with irregularly-positioned buttons. Unlike a regular grid, these buttons are scattered, but auto-navigation figures out the correct connections - also shows currently focused button, and the last "input" pressed to show the logical flow of navigating:

```bash
cargo run --example auto_directional_navigation
```


https://github.com/user-attachments/assets/8dce1d8d-53b5-41b8-bc9a-7e32067f7978




**Key differences from manual `directional_navigation` example**:
- No manual `add_edges()` or `add_looping_edges()` calls
- Buttons positioned irregularly (not in a perfect grid)
- Works with absolute positioning and dynamic layouts

## Migration Guide

No breaking changes - this is a purely additive feature.

To adopt automatic navigation:
1. Add `AutoDirectionalNavigation` component to focusable entities
2. Optionally configure `AutoNavigationConfig` resource

jbuehler23 commented:
> How is the graph connectivity with this algorithm? It's rather suboptimal to have the graph be disconnected, imo worse than oddly skewed and somewhat unintuitive connections.
> 
> Maybe there could be some pattern where we run the solver recursively with increasingly more lenient parameters until the graph is sufficiently dense? (This could be up to the user, but we should have a clean demonstration on how to do it.)
> 
> PS: I would suggest looking at how Shift+Arrow navigation works in a browser as prior art.

With default config, the graph should be weakly connected in most layouts, but it's not guaranteed due to the directional cone filter + overlap requirements.

I do like your progressive relaxation idea, however. We could add a helper function:
```rust
pub fn ensure_connectivity(
    nav_map: &mut DirectionalNavigationMap,
    nodes: &[(Entity, Vec2, Vec2)],
) {
    // Try default config, then relax if disconnected
}
```
Or perhaps better, is to just add a config parameter for the user to define:
```rust
pub struct AutoNavigationConfig {
    // ... existing fields ...

    /// If true, ensures all nodes are reachable by relaxing constraints
    /// for nodes that would otherwise be isolated
    pub ensure_connectivity: bool,
}
```
Then we can demonstrate this usage in the `auto_directional_navigation` example.

I think either of these options give the user control of how to manage the connectivity issue you've described, but remains simple enough so as to not pollute with too many options.   What do you think of this?

Re: browser shift+arrow - based on my research it appears to use projection-based selection to guarantee connectivity. It might be best to add this as an alternative mode in the future, and support other algorithms? Something like:

```rust
pub enum NavigationAlgorithm {
      OverlapBased,  // Current design
      ProjectionBased, // Like browser Shift+Arrow?
}
```

PPakalns commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -35,7 +46,100 @@ pub struct DirectionalNavigationPlugin;
+    /// Whether to also consider `TabIndex` for navigation order hints.
+    /// Currently unused but reserved for future functionality.
+    pub respect_tab_order: bool,
+}
```
I think it can be done in future PR when bevy_ui with multiple layers is being used more widely.

Good to hear that currently there is workaround where `auto_generate_navigation_edges` can be called manually.

jbuehler23 commented:
> It looks like your PR has been selected for a highlight in the next release blog post, but **you didn't provide a release note**.
> 
> Please review the [instructions for writing release notes](https://github.com/bevyengine/bevy/tree/main/release-content/release_notes.md), then expand or revise the content in the [release notes directory](https://github.com/bevyengine/bevy/tree/main/release-content/release-notes) to showcase your changes.

added release notes @alice-i-cecile :)

rossleonardy commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -35,7 +46,100 @@ pub struct DirectionalNavigationPlugin;
+    ///
+    /// For example, when navigating East/West, nodes must have some vertical overlap.
+    /// A value of 0.0 means any overlap is acceptable, while 1.0 means perfect alignment is required.
+    pub min_alignment_factor: f32,
```
I think this could use more doc comments describing what you've said here

rossleonardy commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -259,8 +363,323 @@ pub enum DirectionalNavigationError {
+    derive(Reflect),
+    reflect(Debug, PartialEq, Clone)
+)]
+pub struct NavigationNode {
```
If these are always in a flat list and not a tree, its a little confusing to me to call them nodes. Could they just be `FocusableArea` like in DOM world or NavigationArea, NavigationAnchor, something like that? 
 https://html.spec.whatwg.org/multipage/interaction.html#focusable-area

jbuehler23 replied:
I've refactored to use `FocusableArea`!

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -13,15 +13,26 @@
+//!    and [`add_looping_edges`](DirectionalNavigationMap::add_looping_edges) to explicitly define connections.
+//! 2. **Automatic**: Add the [`AutoDirectionalNavigation`] component to UI entities with
+//!    position and size data. The system will automatically compute nearest neighbors in each direction.
+//!    Manual edges always take precedence over auto-generated ones.
```
This would be clearer as a new paragraph, and should probably call out explicitly that you can combine these approaches.

jbuehler23 replied:
Done in latest :)

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -13,15 +13,26 @@
 //! Each entity can have up to 8 neighbors, one for each [`CompassOctant`], balancing flexibility and required precision.
-//! For now, this graph must be built manually, but in the future, it could be generated automatically.
+//!
+//! # Automatic Navigation Graph Generation
```
```suggestion
//! # Creating a Navigation Graph
```

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -35,7 +46,100 @@ pub struct DirectionalNavigationPlugin;
+        app.init_resource::<DirectionalNavigationMap>()
+            .init_resource::<AutoNavigationConfig>()
+            .add_systems(
+                PostUpdate,
```
The `UiSystems::PostLayout` system set you mean? Yes, I think that's a good idea.

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -35,7 +46,100 @@ pub struct DirectionalNavigationPlugin;
+    derive(Reflect),
+    reflect(Component, Default, Debug, PartialEq, Clone)
+)]
+pub struct AutoDirectionalNavigation {
```
I'm strongly considering suggesting that we add this as a required component to all of our interactable widgets. That can be done in follow-up though.

alice-i-cecile replied:
Hmm 🤔 You could remove the edges, or disable automatic directional navigation computation? Are there use cases where you would want to show users a button but not actually let them navigate to it?

IQuick143 replied:
What if you had for example a pop-up dialog? I imagine you'd want to handle that manually.
Or just generally: UI elements that are obscured, but not flat-out hidden/invisible.

I could also imagine a "defunct" button which is intentionally left unclickable. (For example to communicate the player is unable/forbidden from doing something; I remember this being used as  a narrative device in some games.) The button would probably still be a (disabled) widget, but wouldn't necessarily need navigation.

viridia replied:
More generally, I would like to explore the topic of generalizing sequential and spatial navigation. Obviously, the two modes are going to share some implementation in common: it would be a mistake to try and implement and maintain two completely separate navigation systems. At the same time, they are different modalities: sequential navigation works best on a traditional keyboard with a tab key, while spatial navigation makes sense for a game controller with directional buttons. The key question in my mind is, how much can be shared between these two modes?

I would like it to be easy for widget authors to add support for either sequential or spatial navigation. This should be doable if we take the approach that both navigators operate on a common widget graph, but where the actual navigation behavior is a plugin.

jbuehler23 replied:
Might be worth continuing this conversation on Issue: #21679, and I can update accordingly

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -259,8 +363,323 @@ pub enum DirectionalNavigationError {
+
+/// Convert `CompassOctant` to a unit direction vector.
+///
+fn octant_to_direction(octant: CompassOctant) -> Vec2 {
```
This doesn't need to exist: you should use the `From<CompassOctant> for Dir2` that already exists.

alice-i-cecile commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -259,8 +363,323 @@ pub enum DirectionalNavigationError {
+///
+/// This system only works with `bevy_ui` nodes. For custom UI systems, call
+/// [`auto_generate_navigation_edges`] directly in your own system.
+fn auto_rebuild_ui_navigation_graph(
```
This system needs to account for visibility, and not suggest that users navigate to nodes that are currently set as invisible.

alice-i-cecile reviewed (changes requested):
Did a second pass: I have a few more substantive critiques now too!

jbuehler23 commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -13,15 +13,26 @@
+//!
+//! 1. **Manual**: Use methods like [`add_edge`](DirectionalNavigationMap::add_edge)
+//!    and [`add_looping_edges`](DirectionalNavigationMap::add_looping_edges) to explicitly define connections.
+//! 2. **Automatic**: Add the [`AutoDirectionalNavigation`] component to UI entities with
```
Agreed, done!

jbuehler23 commented on examples/ui/auto_directional_navigation.rs:
```diff
@@ -0,0 +1,471 @@
+
+fn main() {
+    App::new()
+        // Input focus is not enabled by default, so we need to add the corresponding plugins
```
Updated follow-up issue: https://github.com/bevyengine/bevy/issues/21679#issue-3561775791

ickshonpe commented:
> Looking at the example, if you press down to navigate from button 3 to button 5 and then press up, you don't return to button 3 but instead go to button 2. Ideally there should be a history so that navigation can backtrack following the user's route.

Oh I see, button 2 can only have one edge for the upwards direction in the navigation map, so it's not that easy.

ickshonpe commented on crates/bevy_input_focus/src/directional_navigation.rs:
```diff
@@ -259,8 +469,317 @@ pub enum DirectionalNavigationError {
+            }
+        }
+        // Diagonal directions don't require strict overlap
+        _ => 1.0,
```
For diagonals we could project onto the diagonal axis and the compare the overlap, or is that not necessary?

ickshonpe commented on crates/bevy_math/src/compass.rs:
```diff
@@ -170,6 +214,53 @@ impl CompassOctant {
+    /// must have a positive dot product with the direction vector.
+    ///
+    /// **Note**: This function assumes **UI coordinates** where Y increases downward.
+    /// This is the opposite of mathematical coordinates where Y increases upward.
```
I think reversing the y-axis direction for some of the methods of `CompassOctant` is going to be really confusing. Just adding a doc-comment isn't sufficient, expecting the directions to be consistant is very natural.

jbuehler23 commented:
@ickshonpe I think I've addressed all of your review comments now, would appreciate another quick check whenever you can please! Thanks again for going through it!

ickshonpe reviewed (approved):
Happy with this now, looks good.
